## Modeling discount solved for time with a given factor that describes the value X at a time t in the future

In [24]:
import numpy as np

"""
Given a Price To Sales of 12 - 
How many years is a company priced in the future when it's revenue grows by 10 % yearly
"""

# given Conditions
ps = 21 # price to sales
current_revenue = 4.5
g = 0.18 # annual growth rate

$$
S_0 \cdot \beta = S_0 \cdot (1 + g)^t           
$$ <br><br>
$$
\beta = (1 + g)^t             
$$ <br><br>
$$
log(\beta) = t \cdot log(1 + g)           
$$ <br><br>
$$
t = \frac{log(\beta)}{log(1 + g)} 
$$

In [25]:
t = np.log(ps) / np.log(1 + g)
t

18.394301220651187

Lets simulate this process now as a function f(x), taking an interval $[g_b; g_a]$ in it's growth rate as an input, calculating the average time needed for this growth interval.
An Interval describing the annual growth rate fits the real world revenue growth deviations better and gives a better approximation for the time premium payed.

In [26]:
def calculate_time_premium_payed(beta, g_lower, g_upper=None, name=None):
    if g_upper is not None:
        growth_rates = np.arange(g_lower, g_upper + 1, 1)
        
        outputs = []
        
        for g in growth_rates:
            t = np.log(beta) / np.log(1 + g)
            outputs.append(t)
            
        t_mean = np.mean(outputs)
        if t_mean < 0:
            t_mean = np.NaN
        
        print(f"The revenue of {name if name is not None else 'Company X'} trades at a premium of {np.round(t_mean):.0f} years forward at a average growth rate of {np.mean([g_lower, g_upper]) * 100:.2f} %.")
        return t_mean
    
    t = np.log(beta) / np.log(1 + g_lower)
    if t < 0:
        t = np.NaN
        
    print(f"The revenue of {name if name is not None else 'Company X'} trades at a premium of {np.round(t):.0f} years forward at a growth rate of {g_lower * 100:.1f} %.")
    return t

In [27]:
f_x = calculate_time_premium_payed(5, 0.04, 0.08)

The revenue of Company X trades at a premium of 22 years forward at a average growth rate of 6.00 %.


<details>
In this demonstration the value $S_0$ does not affect the outcome in it's intended way.
Furthermore the multiple $\beta$ is dependend on $S_0$.
However this doensn't invalidate the model – it's more a example data problem.
The following real world example demonstrates the model like expected.
</details>

Now let's fetch stock data from the yahoo finance api and implement the model for real world usage

In [28]:
import yfinance as yf

# Note: You could also use yf.Ticker(...).info["priceToSalesTrailing12Months"]

def calculate_price_to_sales(symbol):
    ticker = yf.Ticker(symbol)
    
    name = ticker.info["longName"]
    market_cap = ticker.info["marketCap"]
    revenue = ticker.get_income_stmt(freq="quarterly").loc["TotalRevenue"].sum() # yfinance returns a DataFrame with 4 quarters lookback period
    
    ps = market_cap / revenue
    
    print(f"{name} has a PriceToSalesTTM of {ps:.2f}")
    
    return ps, name

In [29]:
if __name__ == "__main__":
    symbol = "NVDA"
    
    ps, name = calculate_price_to_sales(symbol)
    g_lower = 0.14
    g_upper = 0.22
    
    calculate_time_premium_payed(beta=ps, g_lower=g_lower, g_upper=g_upper, name=name)

NVIDIA Corporation has a PriceToSalesTTM of 20.94
The revenue of NVIDIA Corporation trades at a premium of 14 years forward at a average growth rate of 18.00 %.


Because the solution at the top of the notebook and last one right here differ a bit from each other, there must be a strong sensitivity to changes in the input, that the time at a growth rate of 18% is not equal to the average time of a growth rate confidence interval with a mean of 18%.
This underlines why earnings reactions on big surprises have such significant effect on the traded share price.
A growth rate condidence interval is the first step of approximating the models sensitivity and putting it into a more stable framework.